provider/cluster/hostname.go

```diff
@@ -4,92 +4,227 @@ import (
 	"context"
 	"fmt"
 	lifecycle "github.com/boz/go-lifecycle"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
 	dtypes "github.com/ovrclk/akash/x/deployment/types"
 	"github.com/pkg/errors"
 	"strings"
 	"sync"
 )
 
-type reserveRequest struct {
-	hostnames []string
-	result    chan<- error
-	doReserve bool
-	dID       dtypes.DeploymentID
-}
-
 type HostnameServiceClient interface {
-	ReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error
-	ReleaseHostnames(hostnames []string)
-	CanReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error
+	ReserveHostnames(hostnames []string, dID dtypes.DeploymentID) ReservationResult
+	ReleaseHostnames(dID dtypes.DeploymentID)
+	CanReserveHostnames(hostnames []string, ownerAddr sdktypes.Address) <-chan error
+	PrepareHostnamesForTransfer(hostnames []string, dID dtypes.DeploymentID) <- chan error
 }
 
 type SimpleHostnames struct {
 	Hostnames map[string]dtypes.DeploymentID
 	lock      sync.Mutex
 } /* Used in test code */
 
-func (sh *SimpleHostnames) ReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error {
+type ReservationResult struct {
+	ChErr               <-chan error
+	ChWithheldHostnames <-chan []string
+}
+
+func (rr ReservationResult) Wait(wait <-chan struct{}) ([]string, error) {
+	select {
+	case err := <-rr.ChErr:
+		return nil, err
+	case v := <-rr.ChWithheldHostnames:
+		return v, nil
+	case <-wait:
+		return nil, errors.New("bob")
+	}
+}
+
+func (sh *SimpleHostnames) PrepareHostnameForTransfer(hostnames []string, dID dtypes.DeploymentID) <- chan error{
 	sh.lock.Lock()
 	defer sh.lock.Unlock()
-	ch := make(chan error, 1)
+	errCh := make(chan error, 1)
+	prepareHostnamesImpl(sh.Hostnames, hostnames, dID, errCh)
+	return errCh
+}
+
+func prepareHostnamesImpl(store map[string]dtypes.DeploymentID, hostnames []string, dID dtypes.DeploymentID, errCh chan <- error){
+	ownerAddr, err := dID.GetOwnerAddress()
+	if err != nil {
+		errCh <- err
+		return
+	}
+	toChange := make([]string, 0, len(hostnames))
 	for _, hostname := range hostnames {
-		_, inUse := sh.Hostnames[hostname]
+		existingDID, ok := store[hostname]
+		if ok {
+			existingOwnerAddr, err := existingDID.GetOwnerAddress()
+			if err != nil {
+				errCh <- err
+				return
+			}
+			if existingOwnerAddr.Equals(ownerAddr) {
+				toChange = append(toChange, hostname)
+			} else {
+				errCh <- fmt.Errorf("%w: host %q in use", ErrHostnameNotAllowed, hostname)
+			}
+		}
+	}
+
+	for _, hostname := range toChange {
+		store[hostname] = dID
+	}
+	close(errCh)
+}
+
+func (sh *SimpleHostnames) ReserveHostnames(hostnames []string, dID dtypes.DeploymentID) ReservationResult {
+	sh.lock.Lock()
+	defer sh.lock.Unlock()
+	errCh := make(chan error, 1)
+	resultCh := make(chan []string, 1)
```

Comment:
It doesn't look like any threads are used here.  Why do we need channels?

Reply:
This type is used in tests to avoid the need to stop / start another goroutine in each test, so it uses the same code path as the normal one as well.

Reply:
can you put these in a file name that ends in `_test.go`?

Reply:
its used in other packages like in the file `provider/manifest/manager_test.go`. Is it still import-able if the definition is in `_test.go` file?

Reply:
I believe so, but I could be wrong.

Reply:
if I move this type into a file ending in `_test.go` then try and run that test I get the following error

```
~/ovrclk/akash/provider/manifest$ go test -v ''
# github.com/ovrclk/akash/provider/manifest [github.com/ovrclk/akash/provider/manifest.test]
./manager_test.go:84:16: undefined: "github.com/ovrclk/akash/provider/cluster".SimpleHostnames
```